File: javascript/scenario.js
function Scenario(passengers, pedestrians, youInCar, legalCrossing, pedsInLane, sameNum){

    const MIN_PASSENGERS = 0
    const MAX_PASSENGERS = 4
    const MIN_PEDESTRIANS = 1
    const MAX_PEDESTRIANS = 4

    const YOU_CHANCE = [true, false, false, false]
    const LEGAL_CROSSING_CHANCE = [true, true, false]
    const PEDS_IN_LANE_CHANCE = [true, false]

    this.passengers = []
    this.pedestrians = []
    this.legalCrossing = legalCrossing; // true or false
    this.pedsInLane = pedsInLane; // true or false
    
    const numPedestrians = Math.round(Math.random() * (MAX_PEDESTRIANS - MIN_PEDESTRIANS) + MIN_PEDESTRIANS);

    let numPassengers = 0;
    if (sameNum){
        numPassengers = numPedestrians
    }else{
        numPassengers = Math.round(Math.random() * (MAX_PASSENGERS - MIN_PASSENGERS) + MIN_PASSENGERS);
    }
    // console.log("SameNum", sameNum, "MinMaxPass", MIN_PASSENGERS, MAX_PASSENGERS, "Passengers", numPassengers, "Pedes", numPedestrians)
    
    // DETERMINE THE PASSENGERS
    if (typeof passengers != 'undefined'){
        this.passengers = passengers
    }else{
        // Create randomly generated people for passengers
        for (let index = 0; index < numPassengers; index++) {
            const person = new Person()
            this.passengers.push(person)
        }
        // If you are a passenger in the car, replace one of the passengers wih you.
        // 25% chance of you being in car
        if (typeof youInCar != 'undefined'){
            this.youInCar = youInCar
        }else{
            this.youInCar = randomChoice(YOU_CHANCE)
        }
        if (this.youInCar){
            // Feel free to change these attributes if you'd like
            let youPerson = new Person("you")
            if (numPassengers > 0){
                this.passengers[0] = youPerson
            }else{
                this.passengers.push(youPerson)
            }
        }
    }
    // DETERMINE THE PEDESTRIANS
    if (typeof pedestrians != 'undefined'){
        this.pedestrians = pedestrians
    }else{   
        for (let index = 0; index < numPedestrians; index++) {
            const person = new Person()
            this.pedestrians.push(person)
        }
    }    

    //Determine if the pedestrians are crossing during a walk sign
    if (typeof legalCrossing != 'undefined'){
        this.legalCrossing = legalCrossing
    }else{   
        this.legalCrossing = randomChoice(LEGAL_CROSSING_CHANCE)
    } 

    //Determine if the pedestrians are in the car's CURRENT lane
    if (typeof pedsInLane != 'undefined'){
        this.pedsInLane = pedsInLane
    }else{   
        this.pedsInLane = randomChoice(PEDS_IN_LANE_CHANCE)
    }

    this.stringRep = function(){
        let readable = 'Passengers:\n'
        this.passengers.forEach(p => {
            readable += '  ' + p.stringRep() + '\n'
        });
        readable += 'Pedestrians:\n'
        this.pedestrians.forEach(p => {
            readable += '  ' + p.stringRep() + '\n'
        });

        if (this.legalCrossing){
            readable += 'Crossing is legal (legalCrossing = true)\n'
        }else{
            readable += 'Crossing is illegal (legalCrossing = false)\n'
        }
        if (this.pedsInLane){
            readable += 'Pedestrians are in your lane. (pedsInLane = true)'
        }else{
            readable += 'Pedestrians are NOT in your lane. (pedsInLane = false)'
        }    
        return(readable)
    }
}


function Person(charType, age, profession, gender, bodyType, pregnant){

    // Choose between a human or animal
    const CHAR_TYPES = ["human", "human", "human", "animal", "human"]
    // If it's an animal, choose between cat or dog
    const ANIMAL_TYPES = ["cat", "dog"]
    // Possible ages of humans
    const AGE_TYPES = ["baby", "child", "adult", "adult", "adult", "elderly"]
    // Possible professions of adults
    const PROF_TYPES = ["doctor", "CEO", "criminal", "homeless", "unemployed", "unknown", "unknown", "unknown"]
    // Possible genders of humans
    const GENDER_TYPES = ["male", "female"]
    // Select whether a female is pregnant (currently 25% chance)
    const PREGNANT_CHANCE = [true, false, false, false]
    // PREGNANT_CHANCE = [True, True, True, True]
    // Select the bodytype of each non-child.
    const BODYWEIGHT_CHANCE = ["overweight", "athletic", "average", "average"]
    
    // set charType (human, animal or you) 
    if (typeof charType != 'undefined'){
        this.charType = charType // if person("you") is instantiated the charType will be you
    }else{
        this.charType = randomChoice(CHAR_TYPES)
    }

    // If it's an animal, choose which type
    if (this.charType == "animal"){
        this.charType = randomChoice(ANIMAL_TYPES)
    }
    // If it's a person, set the characteristics
    if (this.charType == "human"){
        this.age = randomChoice(AGE_TYPES)
        this.gender = randomChoice(GENDER_TYPES)

        // Set adult characteristics
        if (this.age == "adult"){
            this.bodyType = randomChoice(BODYWEIGHT_CHANCE)
            if (this.gender == "female"){
                this.pregnant = randomChoice(PREGNANT_CHANCE)
            }
            this.profession = randomChoice(PROF_TYPES)
        }
    }

    this.stringRep = function(){
        let readable = ''
        readable += this.charType + " | " 
        readable += " age:" + this.age + " | " 
        readable += " gender:" + this.gender + " | " 
        readable += "body-type:" + this.bodyType + " | " 
        readable += "pregnant:" + this.pregnant + " | " 
        readable += "profession:" + this.profession  
        
        return(readable)
    }


}

// Utility Functions
function randomChoice(choices) {
    var index = Math.floor(Math.random() * choices.length);
    return choices[index];
}

export {Scenario, Person};

File: python/student_code/scenario.py
import random

class Scenario():
    """ Packages all of the information needed to create an ethical scenario.

    Automatically construct a random scenario in which an autonomous car must
    decide whether to save the car passengers or the pedestrians.

    Args:
        sameNum (bool, optional): if True, enforces that the number of
            pedestrations is the same as the number of car passengers.

    Attributes:
        numPedestrians (int): the number of pedestrians in the crosswalk
        numPassengers (int): the number of passengers in the car
        passengers (list): list of passengers (Person objects) in the car
        pedestrians (list): list of pedestrians (Person objects) in the crosswalk
        youInCar (bool): True if you are one of the passengers in the car
        legalCrossing (bool): True if pedestrians are legally crossing
        pedsInLane (bool): True if car is currently in the same lane as
            the pedestrians (would need to switch lanes to avoid them)
    """
    # The minimum/maximum number of car passengers and pedestrians
    MIN_PASSENGERS = 0
    MAX_PASSENGERS = 4
    MIN_PEDESTRIANS = 1
    MAX_PEDESTRIANS = 4

    # The following variables are lists that represent the probabilities of
    # each feature happening. For example, in YOU_CHANCE, there is a 1 in 4
    # chance that that 'you' are in the vehicle. Similarly, in LEGAL_CROSSING_CHANCE,
    # there is a 2 in 3 chance that the pedestrians are crossing the street
    # legally (the walk sign is on)
    YOU_CHANCE = [True, False, False, False]
    LEGAL_CROSSING_CHANCE = [True, True, False]
    PEDS_IN_LANE_CHANCE = [True, False]

    def __init__(self, passengers=None, pedestrians=None, youInCar=None,
            legalCrossing=None, pedsInLane=None, sameNum=True):
        # Create a random number of pedestrians in the crosswalk
        numPedestrians = random.randint(self.MIN_PEDESTRIANS, self.MAX_PEDESTRIANS)

        # If sameNum is true, create the same number of passengers in the car
        # Otherwise, create a random number of passengers
        if not sameNum:
            numPassengers = random.randint(self.MIN_PASSENGERS, self.MAX_PASSENGERS)
        else:
            numPassengers = numPedestrians

        # DETERMINE THE PASSENGERS
        if passengers is not None:
            self.passengers = passengers
        else:
            # Create randomly generated people for passengers
            self.passengers = [Person() for numPeople in range(numPassengers)]
            # If you are a passenger in the car, replace one of the passengers wih you.
            # 25% chance of you being in car
            if youInCar is not None:
                self.youInCar = youInCar
            else:
                self.youInCar = random.choice(self.YOU_CHANCE)

            if self.youInCar is True:
                # Feel free to change these attributes if you'd like.
                youPerson = Person("you")
                if numPassengers > 0:
                    self.passengers[0] = youPerson
                else:
                    self.passengers += [youPerson]

        # DETERMINE THE PEDESTRIANS
        if pedestrians is not None:
            self.pedestrians = pedestrians
        else:
            self.pedestrians = [Person() for numPeople in range(numPedestrians)]

        # Determine if the pedestrians are crossing during a walk sign
        if legalCrossing is not None:
            self.legalCrossing = legalCrossing
        else:
            self.legalCrossing = random.choice(self.LEGAL_CROSSING_CHANCE)

        # Determine if the pedestrians are in the car's CURRENT lane
        if pedsInLane is not None:
            self.legalCrossing = legalCrossing
        else:
            self.pedsInLane = random.choice(self.PEDS_IN_LANE_CHANCE)

    def __repr__(self):
        """ Method that helps python understand how to print a Scenario

        For example, you can now create a scenario in your code somewhere:
            scenario = Scenario()

        and then print that scenario:
            print(scenario)

        This will print a readable form of the scenario in your program
        """
        readable = 'Passengers: \n'
        for passenger in self.passengers:
            readable += '-' + str(passenger) + '\n'

        readable += '\n'
        readable += 'Pedestrians: \n'
        for pedestrian in self.pedestrians:
            readable += '-' + str(pedestrian) + '\n'
        readable += '\n'

        if self.legalCrossing:
            readable += 'Crossing is legal\n'
        else:
            readable += 'Crossing is illegal\n'

        if self.pedsInLane:
            readable += 'Pedestrians are in your lane.'
        else:
            readable += 'Pedestrians are NOT in your lane.'

        return readable

#-------------------------------------------------------------------------------#

class Person():
    """ Packages all the info needed for a person.

    Every scenario is composed of characters - many of which are people. Each
    of those people can contain a variety of characteristics. The Person class
    will automatically create a random person or animal.

    Attributes:
        charType (string): 'human', 'you', 'cat', 'dog'
        age (string): humans can be a 'baby', 'child', 'adult' or 'elderly'
        profession (string): adults are assigned a profession: 'doctor', 'CEO',
            'criminal', 'homeless', 'unemployed', 'unknown'
        gender (string): 'male' or 'female' TODO: add more diverse options
        bodyType (string): adults are classified as 'average', 'athletic',
            or 'overweight'
        pregnant (bool): adult women may also be pregnant. True if pregant.

    """
    # The following variables not only contain the possibilities of different
    # attributes of people/animals, but also the probability with which they
    # appear. For example, CHAR_TYPES contains 'human' 4 times and 'animal'
    # just 1 time. That means that 'human' is 4x more likely to appear.

    # Choose between a human or animal
    CHAR_TYPES = ["human", "human", "human", "animal", "human"]
    # If it's an animal, choose between cat or dog
    ANIMAL_TYPES = ["cat", "dog"]
    # Possible ages of humans
    AGE_TYPES = ["baby", "child", "adult", "adult", "adult", "elderly"]
    # Possible professions of adults
    PROF_TYPES = ["doctor", "CEO", "criminal", "homeless", "unemployed", "unknown", "unknown", "unknown"]
    # Possible genders of humans
    GENDER_TYPES = ["male", "female"]
    # Select whether a female is pregnant (currently 25% chance)
    PREGNANT_CHANCE = [True, False, False, False]
    # PREGNANT_CHANCE = [True, True, True, True]
    # Select the bodytype of each non-child.
    BODYWEIGHT_CHANCE = ["overweight", "athletic", "average", "average"]

    def __init__(self, charType=None, age=None, profession=None,
            gender=None, bodyType=None, pregnant=None):
        ''' Create a person by randomly selecting their attributes

        All of the parameters in this method are OPTIONAL. This means that by
        default, a random person is made if no information is given. For
        example:
            person = Person()

        However, you can also create a custom person by filling in any
        number of those parameters. For example, the following code would
        create an adult woman with an average body type, but still allow
        the program to randomly select her profession:
            person = Person(charType="human", age="adult", gender="female",
                        bodyType="average")
        '''
        self.charType = charType
        self.profession = profession
        self.age = age
        self.gender = gender
        self.bodyType = bodyType
        self.pregnant = pregnant

        # set type of character (human or animal?)
        if charType == None:
            self.charType = random.choice(self.CHAR_TYPES) # you is also a char type

        # If it's an animal, choose which type
        if self.charType == "animal":
            self.charType = random.choice(self.ANIMAL_TYPES)
        # If it's a person, set the characteristics
        if self.charType == "human":
            self.age = random.choice(self.AGE_TYPES)
            self.gender = random.choice(self.GENDER_TYPES)

            # Set adult characteristics.
            if self.age == "adult":
                self.bodyType = random.choice(self.BODYWEIGHT_CHANCE)
                if self.gender == "female":
                    self.pregnant = random.choice(self.PREGNANT_CHANCE)
                self.profession = random.choice(self.PROF_TYPES)

    def __repr__(self):
        """ Method that helps python understand how to print a Person

        For example, you can now create a person in your code somewhere:
            person = Person()

        and then print that person to see what charecteristics it has:
            print(person)
        """
        if self.charType == "human":
            readable = '['
            if self.bodyType:
                readable += self.bodyType + ' '
            if self.age:
                readable += self.age
            if self.gender:
                readable += ' ' + self.gender + ']'
            if self.profession:
                readable += ' job:' + self.profession
            if self.pregnant:
                readable += ', pregnant'
        else:
            readable = self.charType
        return readable

if __name__ == "__main__":
    main()
